src/hp/derp/http/server.rs

```diff
@@ -1,50 +1,404 @@
+        } else {
+            (DerpHandler::Override(self.derp_override.unwrap()), None)
+        };
+        let h = self.headers.clone();
+        let not_found_fn = match self.not_found_fn {
+            Some(f) => f,
+            None => Box::new(move |mut res: ResponseBuilder| {
+                for (k, v) in h.iter() {
+                    res = res.header(*k, *v);
+                }
+                let r = res
+                    .status(StatusCode::NOT_FOUND)
+                    .body(b"Not Found"[..].into())
+                    .unwrap();
+                HyperResult::Ok(r)
+            }),
+        };
+
+        let service = DerpService::new(
+            self.handlers,
+            derp_handler,
+            self.derp_endpoint,
+            not_found_fn,
+            self.headers,
+        );
+
+        let server_state = ServerState {
+            addr: self.addr,
+            tls_config: self.tls_config,
+            server: derp_server,
+            service,
+        };
+
+        server_state.serve().await
+    }
+}
+
+impl std::fmt::Debug for ServerBuilder {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let hyper_fn =
+            "Some(Box<dyn Fn(ResponseBuilder) -> HyperResult<Response<Body>> + Send + Sync + 'static)";
+        let derp_override = if let Some(_) = self.derp_override {
+            hyper_fn
+        } else {
+            "None"
+        };
+
+        let not_found_fn = if let Some(_) = self.not_found_fn {
+            hyper_fn
+        } else {
+            "None"
+        };
 
-    // send to the derp server
-    conn_handler.accept(reader, writer).await
+        write!(f, "ServerBuilder {{ secret_key: {:?}, addr: {:?}, mesh_key: {:?}, tls_config: {:?}, handlers: {:?}, derp_endpoint: {:?}, derp_override: {derp_override}, headers: {:?}, not_found_fn: {not_found_fn}  }}", self.secret_key, self.addr, self.mesh_key, self.tls_config, self.handlers, self.derp_endpoint, self.headers)
+    }
+}
+
+#[derive(Debug)]
+pub struct ServerState {
+    addr: SocketAddr,
+    tls_config: Option<TlsConfig>,
+    server: Option<crate::hp::derp::server::Server<HttpClient>>,
+    service: DerpService,
 }
 
-impl<P> hyper::service::Service<Request<Body>>
-    for ClientConnHandler<OwnedReadHalf, OwnedWriteHalf, P>
+impl ServerState {
+    async fn serve(self) -> Result<Server> {
+        match &self.tls_config {
+            Some(_) => self.https_serve().await,
+            None => self.http_serve().await,
+        }
+    }
+
+    /// Binds a TCP listener on `addr` and handles content using HTTP.
+    /// Returns the local `SocketAddr` on which the server is listening.
+    async fn http_serve(self) -> Result<Server> {
+        let http_listener = TcpListener::bind(&self.addr)
+            .await
+            .context("failed to bind https")?;
+        let addr = http_listener.local_addr()?;
+        let cancel = CancellationToken::new();
+        info!("[HTTP] derp: serving on {addr}");
```

Comment:
If you'd like to be fancier you can like do `self.http_serve().instrument(debug_span!("derp_http")).await` on line 296 and then use `info!(%addr, "serving");` here.